feat: Upgrade guide for Vaadin 24.4 (#3249)
vaadin/docs: articles/flow/upgrading/index.adoc

As an alternative, you can set `bean-discovery-mode=all` in the [filename]`beans.xml` file if it's applicable to your project. However, this is not recommended.

[role="since:com.vaadin:vaadin@V24.4"]
== Unification With Hilla And Integration With React

> Vaadin 24 unifies the server and client development approaches known as Vaadin Flow and Hilla respectively.

Reviewer: Should this be specifically `24.4+`

Author: Added

> Vaadin 24 unifies the server and client development approaches known as Vaadin Flow and Hilla respectively.

Author: ```suggestion
Vaadin 24.4 and later versions unify the server and client development approaches known as Vaadin Flow and Hilla respectively.
```

It makes easier for developers to create hybrid applications, including:

- unified dependencies for entire Vaadin platform, making the project configuration simpler;

> - utilisation of the https://reactjs.org/[React] for making UI in Vaadin Flow applications;

Reviewer 2: ```suggestion
- develop https://reactjs.org/[React] based Routes or Components in Vaadin Flow applications;
```
Just a suggestion for better marketing ;)

Author: Thanks, applied 👍

- export Flow components or entire views into frontend views written in React;
- use unified routing based on https://reactrouter.com/en/main[React Router];

=== Hilla And React Dependencies

> Vaadin 24 includes React dependencies, such as React core, router, and React components provided by Vaadin, that allows starting React development immediately.

Reviewer 2: > Vaadin 24 includes React dependencies, such as React(link?), React Router (link?), and React based Components provided by Vaadin (link?), that allows starting React development immediately.

Capitalized / Wording the react things.. Should there be links for people to follow up?

Author: Added

Vaadin's React components bring all Pro Web Components dependencies in [filename]`node_modules`, although it will not be a part of the frontend bundle, if not used in the project.

Vaadin Flow applications don't need any changes regarding React or Hilla dependencies management, though the opt-out options are available if you would want to avoid these in your project:

- if `Lit` is used for frontend development instead of `React`;
- having Hilla or React dependencies is undesirable;
- use only core, and installing Pro components is undesirable;

> - legacy `vaadin-route` is needed, e.g. if a project has add-on for it.

Reviewer 2: ```suggestion
- legacy `vaadin-router` is needed, e.g. if a project has add-on for it.
```

> Use a `reactEnable` configuration parameter, it excludes the React completely from a project and fallbacks to `vaadin-router` and `Lit`:

Reviewer: ```suggestion
Use the `reactEnable` configuration parameter, it excludes the React completely from a project and fallbacks to `vaadin-router` and `Lit`:
```

> Use the `reactEnable` configuration parameter, it excludes the React completely from a project and fallbacks to `vaadin-router` and `Lit`:

Reviewer 2: ```suggestion
Use the `reactEnable` configuration parameter, it excludes React completely from a project and fallbacks to `vaadin-router` and `Lit`:
```

[.example]
--

[source,properties]
.application.properties
----
vaadin.react.enable=false
----

[source,xml]
.pom.xml
----
<plugin>
   <groupId>com.vaadin</groupId>
   <artifactId>vaadin-maven-plugin</artifactId>
   <configuration>

>        <reactEnable>false</frontendHotdeploy>

Reviewer: ```suggestion
       <reactEnable>false</reactEnable>
```

Author: :)

   </configuration>
</plugin>
----

[source,terminal]
----
-Dvaadin.react.enable=false
----

> Not as an exact alternative, but one can also exclude Vaadin's React components (`@vaadin/react-components` package) and React integration related files from Vaadin Flow applications. This makes sense to do if you use an application's specific frontend bundle and don't want React components to be installed and present in the [filename]`package.json`:

Reviewer 2: Should there be a mention somewhere that frontendHotdeploy is enabled if hilla is found?

Author: Yes, good point. Added

> Not as an exact alternative, but one can also exclude Vaadin's React components (`@vaadin/react-components` package) and React integration related files from Vaadin Flow applications. This makes sense to do if you use an application's specific frontend bundle and don't want React components to be installed and present in the [filename]`package.json`:

Reviewer 2: ```suggestion
If you don't need our React based Component (Wrapper?), you can opt-out by excluding the `flow-react` package in your dependencies. This replaces the `@vaadin/react-components` package in your [filename]`package.json` with `@vaadin/*` web-components.:
```

Not sure which version is better :)

Author: I'm not so good in tech writing, but hope our experts in this area would help us :)

Reviewer 2: Me neither :) Just wanted to make sure this is easy to understand, so that I don't have to write a guide for my users  and can redirect them to the docs :)
